### ISSUE TITLE
Realm Crash on notification listener

### ISSUE THREAD

**Issue by Gao233** [realm/realm-swift#7774]
### How frequently does the bug occur?

All the time

### Description

We are having a hard crash with Realm. With given stack trace, we are having a very hard time debugging it. Would be deeply appreciated if you can give us some guidance! 

This happens only when we lock iPad (just background the app won't appear). We were doubting it's because of iOS's file protection. We even tried to use No protection and got rid of Realm's encryption, crash still exists. 

We would love your opinions on: 
1. If this is a known Realm issue
2. Any misusage of Realm might've caused this issue

Appreciate your time in advance!

### Stacktrace & log output

```shell
Realm notification listener (15)#0	0x0000000109692428 in realm::DB::grab_read_lock(realm::DB::ReadLockInfo&, realm::VersionID) ()
#1	0x00000001091b1058 in bool realm::Transaction::internal_advance_read<realm::_impl::NullInstructionObserver>(realm::_impl::NullInstructionObserver*, realm::VersionID, realm::_impl::History&, bool) at /Users/cheng/Library/Developer/Xcode/DerivedData/Assessments-cnnsdzpbooxcnifzuvcgednnzgcz/Build/Products/Debug-iphoneos/XCFrameworkIntermediates/Realm/Headers/realm/db.hpp:945
#2	0x0000000109230668 in void realm::Transaction::advance_read<realm::_impl::NullInstructionObserver>(realm::_impl::NullInstructionObserver*, realm::VersionID) at /Users/cheng/Library/Developer/Xcode/DerivedData/Assessments-cnnsdzpbooxcnifzuvcgednnzgcz/Build/Products/Debug-iphoneos/XCFrameworkIntermediates/Realm/Headers/realm/db.hpp:849
#3	0x000000010922b658 in realm::Transaction::advance_read(realm::VersionID) at /Users/cheng/Library/Developer/Xcode/DerivedData/Assessments-cnnsdzpbooxcnifzuvcgednnzgcz/Build/Products/Debug-iphoneos/XCFrameworkIntermediates/Realm/Headers/realm/db.hpp:583
#4	0x000000010922b4cc in realm::_impl::ResultsNotifier::do_prepare_handover(realm::Transaction&) at /Users/cheng/repositories/care-at-home-clinical/Pods/Realm/Realm/ObjectStore/src/impl/results_notifier.cpp:172
#5	0x0000000109119b8c in realm::_impl::CollectionNotifier::prepare_handover() at /Users/cheng/repositories/care-at-home-clinical/Pods/Realm/Realm/ObjectStore/src/impl/collection_notifier.cpp:313
#6	0x00000001091d7354 in realm::_impl::RealmCoordinator::run_async_notifiers() at /Users/cheng/repositories/care-at-home-clinical/Pods/Realm/Realm/ObjectStore/src/impl/realm_coordinator.cpp:1075
#7	0x00000001091d6654 in realm::_impl::RealmCoordinator::on_change() at /Users/cheng/repositories/care-at-home-clinical/Pods/Realm/Realm/ObjectStore/src/impl/realm_coordinator.cpp:836
#8	0x000000010913d8c0 in realm::_impl::ExternalCommitHelper::listen() at /Users/cheng/repositories/care-at-home-clinical/Pods/Realm/Realm/ObjectStore/src/impl/apple/external_commit_helper.cpp:231
#9	0x0000000109141fe4 in realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0::operator()() const at /Users/cheng/repositories/care-at-home-clinical/Pods/Realm/Realm/ObjectStore/src/impl/apple/external_commit_helper.cpp:173
#10	0x0000000109141f80 in decltype(static_cast<realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0>(fp)()) std::__1::__invoke<realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0>(realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0&&) at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.4.sdk/usr/include/c++/v1/type_traits:3918
#11	0x0000000109141f1c in void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0>&, std::__1::__tuple_indices<>) at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.4.sdk/usr/include/c++/v1/thread:287
#12	0x0000000109141708 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, realm::_impl::ExternalCommitHelper::ExternalCommitHelper(realm::_impl::RealmCoordinator&)::$_0> >(void*) at /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.4.sdk/usr/include/c++/v1/thread:298
#13	0x00000001f1d3b9a4 in _pthread_start ()
```

### Can you reproduce the bug?

Yes, always

### Reproduction Steps

I can only reproduce in our app. I am trying to get a sample app but so far I have no luck.

### Version

10.25.1

### What SDK flavour are you using?

Local Database only

### Are you using encryption?

Yes, using encryption

### Platform OS and version(s)

iPad 15.3.1

### Build environment

Xcode version: 13.3.1
Dependency manager and version: CocoaPods

**Comment by ericjordanmossman** [comment]
Hi @Gao233 
This looks similar to #6433, which didn't really resolve conclusively.

1. Are you using realms across threads? If so, how?
2. How are you changing the file protection of the realm files? See [here ("background app refresh")](https://www.mongodb.com/docs/realm-legacy/docs/swift/latest/index.html#using-the-realm-framework) and [here](https://github.com/realm/realm-swift/issues/6433#issuecomment-595203764).

**Comment by Gao233** [comment]
Hi @ericjordanmossman , 

Thank you for your response! Yes, I saw #6433 as well and tried couple things in that thread and it didn't end anywhere. 

1. Yes, I am using DispatchQueue(label: "background") thread and main thread for realms. Reason being on main thread is there were some legacy code. (I used PromiseKit and PromiseKit default returns to main thread). But I am pretty confident about we are read/write realm data on the same thread since the crash of accessing from different thread is pretty obvious. 

2. I am following [background app refresh](https://www.mongodb.com/docs/realm-legacy/docs/swift/latest/index.html#using-the-realm-framework) documentation. I am setting **kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly** for **kSecAttrAccessible**

I also wonder if this could be related to` try! realm.write {}`. I had a lot of that in codebase when they first started and just simply followed the base tutorial.
